lib/backend/syncersv1/updateprocessors/bgpconfigprocessor.go

```diff
@@ -100,3 +105,39 @@ var svcClusterIpsToString = func(value interface{}) interface{} {
 
 	return strings.Join(ipCidrs, ",")
 }
+
+// return JSON encoded string of CommunityKVPair
+var communitiesToString = func(value interface{}) interface{} {
```

Comment:
I don't think there's actually any reason we need to convert these to a string. We should be able to send the structured data over as the Value of a KVPair.

Reply:
Same for prefixAdvertisements and listenPort

Reply:
I've updated the communities and prefixAdvertisements to send data as Value of a KVPair. 
I see string is return as it is in many cases, doing the same for listenPort which is just an integer.

Reply:
Looks like we need to serialize it, https://github.com/projectcalico/confd/blob/master/pkg/backends/calico/client.go#L919.

Reply:
That serialization isn't relevant to this code, though, and should work just fine even without this logic. Note the generic json Marshaling at the end: https://github.com/projectcalico/libcalico-go/blob/dfe3bf9034f6399f882ed7c02f98030d472234c8/lib/backend/model/keys.go#L370

Reply:
Returning struct as value of KVPair throws error in confd while serializing (https://github.com/projectcalico/confd/blob/master/pkg/backends/calico/client.go#L919).

Reason seems to be the valueType for GlobalBGPConfigKey is string, so we don't reach the line you pointed above.
https://github.com/projectcalico/libcalico-go/blob/dfe3bf9034f6399f882ed7c02f98030d472234c8/lib/backend/model/keys.go#L354